sdk/storage/Azure.Storage.Common/src/Shared/ClientsideEncryption/ClientSideDecryptor.cs

```diff
@@ -0,0 +1,253 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.IO;
+using System.Security.Cryptography;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core.Cryptography;
+using Azure.Storage.Cryptography.Models;
+
+namespace Azure.Storage.Cryptography
+{
+    internal static class ClientSideDecryptor
+    {
+        /// <summary>
+        /// Decrypts the given stream if decryption information is provided.
+        /// Does not shave off unwanted start/end bytes, but will shave off padding.
+        /// </summary>
+        /// <param name="ciphertext">Stream to decrypt.</param>
+        /// <param name="encryptionData">
+        /// Encryption metadata and wrapped content encryption key.
+        /// </param>
+        /// <param name="ivInStream">
+        /// Whether to use the first block of the stream for the IV instead of the value in
+        /// <paramref name="encryptionData"/>. Generally for partial blob downloads where the
+        /// previous block of the ciphertext is the IV for the next.
+        /// </param>
+        /// <param name="keyResolver">
+        /// Resolver to fetch the key encryption key.
+        /// </param>
+        /// <param name="potentialCachedKeyWrapper">
+        /// Clients that can upload data have a key encryption key stored on them. Checking if
+        /// a cached key exists and matches the <paramref name="encryptionData"/> saves a call
+        /// to the external key resolver implementation when available.
+        /// </param>
+        /// <param name="noPadding">
+        /// Whether to ignore padding. Generally for partial blob downloads where the end of
+        /// the blob (where the padding occurs) was not downloaded.
+        /// </param>
+        /// <param name="async">Whether to perform this function asynchronously.</param>
+        /// <param name="cancellationToken"></param>
+        /// <returns>
+        /// Decrypted plaintext.
+        /// </returns>
+        /// <exception cref="Exception">
+        /// Exceptions thrown based on implementations of <see cref="IKeyEncryptionKey"/> and
+        /// <see cref="IKeyEncryptionKeyResolver"/>.
+        /// </exception>
+        public static async Task<Stream> DecryptInternal(
+            Stream ciphertext,
+            EncryptionData encryptionData,
+            bool ivInStream,
+            IKeyEncryptionKeyResolver keyResolver,
+            IKeyEncryptionKey potentialCachedKeyWrapper,
+            bool noPadding,
+            bool async,
+            CancellationToken cancellationToken)
```

Comment:
I'd rather model this class in same way `BlobClientSideDecryptor` is. `IKeyEncryptionKey` and `IKeyEncryptionKeyResolver` are invariant for `DecryptInternal` so they're more of class dependencies rather than input. `BlobClientSideDecryptor` could take `ClientSideDecryptor` as a dependency.

There are also benefits of killing static code, see https://objcsharp.wordpress.com/2013/07/08/why-static-code-is-bad/ for more detailed explanation. It's a cosmetic change now, can be a headache in long run.

Reply:
Agreed. I believe the package-specific instance decryptors are better. And this would help the possible future where more modes of encryption are added.